calico/felix/test/test_fetcd_api.py

```diff
@@ -0,0 +1,177 @@
+        # instead of writing to etcd.
+        gevent.sleep(1)
+        self.assertFalse(self.etcd_api.write_to_etcd.called)
+
+    def test_update_felix_status_single(self):
+        """
+        Test felix status is updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=15,
+                          REPORTING_TTL_SECS=37)
+
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(1)
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+        self.etcd_api.write_to_etcd.assert_has_calls([status_call,
+                                                      uptime_call])
+
+    def test_update_felix_status_continuous(self):
+        """
+        Test felix status is being continuously updated
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=5,
+                          REPORTING_TTL_SECS=17)
+        hostname = self.etcd_api._config.HOSTNAME
+        status_key = key_for_status(hostname)
+        uptime_key = key_for_uptime(hostname)
+        ttl = self.etcd_api._config.REPORTING_TTL_SECS
+
+        status_call = call(status_key, TestIfStatus(), async=True)
+        uptime_call = call(uptime_key, TestIfUptime(), ttl=ttl, async=True)
+
+        # We call gevent.sleep so that _status_reporting thread is executed
+        gevent.sleep(5)
+
+        last_status_call = None
+        for update in range(5):
+            gevent.sleep(5)
+            # Check that both uptime and status were updated
+            self.etcd_api.write_to_etcd.assert_has_calls([status_call, uptime_call])
+            # Update last_status_call and reset write_to_etcd calls
+            last_status_call = self.etcd_api.write_to_etcd.mock_calls[0]
+            self.etcd_api.write_to_etcd.reset_mock()
+
+    @patch('calico.felix.fetcd._reconnect')
+    def test_update_felix_status_reconnects_on_etcd_exception(self, _reconnect):
+        """
+        Test felix status handles exceptions
+        """
+        self.finish_setup(REPORTING_INTERVAL_SECS=4,
+                          REPORTING_TTL_SECS=12)
+        self.etcd_api.write_to_etcd = Mock(side_effect=EtcdException)
+
+        gevent.sleep(1)
+
+        self.assertTrue(_reconnect.called)
+
+
+class TestIfStatus(object):
+    """
+    Used to check whether status has JSON format and contains expected
+    attributes, whis are:
+    - timestamp in ISO 8601 Zulu format
+    """
+    def __eq__(self, other):
+        import re
+        timestamp_regex = re.compile('.*"status_time": "\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z?".*')
+        has_timestamp =  timestamp_regex.match(str(other)) is not None

```

Comment:
Nitpick: lose the extra space.

Separately: I feel like I've commented on this block before (but perhaps I just thought it and never wrote it down). Using a regex to check that a block is JSON formatted and a timestamp is prone to errors. It would be better to:

* Try to parse as JSON (if it throws a parsing error => incorrectly formatted)
* Try to get the `status_time` key (if it doesn't exist => incorrectly formatted)
* Use the `strptime()` method from the datetime module to try to parse the time (if it doesn't work => incorrectly formatted)

A regex is fragile for this task. Some examples:

```python
'<"status_time": "2012-07-04T09:13:26Z>'  # not valid JSON, but will pass this regex
'{"status_time":"2012-07-04T09:13:26Z}'   # valid JSON, but will fail this regex because it's minified
'{"status_time":"9999-99-99T00:00:00Z}'   # valid JSON, will pass the regex, but an impossible timestamp
```

I'm also not clear why this isn't a function, like `is_json` below.